tuf/repository_lib.py

```diff
@@ -1416,12 +1430,30 @@ def generate_snapshot_metadata(metadata_directory, version, expiration_date,
   securesystemslib.formats.PATH_SCHEMA.check_match(targets_filename)
   securesystemslib.formats.BOOLEAN_SCHEMA.check_match(consistent_snapshot)
   securesystemslib.formats.NAME_SCHEMA.check_match(repository_name)
+  securesystemslib.formats.BOOLEAN_SCHEMA.check_match(use_length)
+  securesystemslib.formats.BOOLEAN_SCHEMA.check_match(use_hashes)
 
   # Snapshot's 'fileinfodict' shall contain the version number of Root,
   # Targets, and all delegated roles fo the repository.
   fileinfodict = {}
-  fileinfodict[TARGETS_FILENAME] = get_metadata_versioninfo(targets_filename,
-      repository_name)
+
+  length, hashes = securesystemslib.util.get_file_details(
+    os.path.join(metadata_directory, targets_filename),
+    tuf.settings.FILE_HASH_ALGORITHMS)
+
+  length = (use_length and length) or None
+  hashes = (use_hashes and hashes) or None
+
+  targets_role = targets_filename[:-len(METADATA_EXTENSION)]
+
+  targets_file_version = get_metadata_versioninfo(targets_role,
+    repository_name)
+
+  # Make file info dictionary with make_metadata_fileinfo because as of
+  # version 1.0.1 of the tuf spec length and hashes are optional for all
+  # METAFILES in snapshot.json including the top-level targets file.
```

Comment:
As with the docstring, I'm not sure it makes sense to reference a version of the spec here

Reply:
Ahh... I just forgot to update that one. I will do that in a minute.